### PR TITLE
Doing a Rust only overhaul

### LINKED ISSUE #4: Exclude buildings in specific areas
landuse = allotments
landuse = commercial
landuse = industrial

### LINKED ISSUE #5: Create interface usable for OpenLinePlanner
This should include:
pdf: OsmPbfReader<File>
inhabitants: &64
centroid: &bool

### LINKED ISSUE #7: Add population function by known inhabitant distribution.
image
d_rel

rel_a

This might even help to validate the output.

### PR DESCRIPTION
**Pull request by TheNewCivilian**
It is time for a makeover. This time getting rid of the python interface.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            Self::IOError(err) => write!(f, "io error occured: {}", err),
+            _ => write!(f, "some error occured"),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+#[derive(Clone, serde::Deserialize)]
+pub struct Config {
```
Comment (zandemax) [inline src/lib.rs]: Here we could implement the Builder Pattern to construct the Config, and use the config in the public interface so it can be tweaked by the caller of the library.

File: src/parser/housenumber.rs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        match self {
+            ParseError::NotFullyConsumedError(input, rest) => {
+                write!(f, "Parsing failed for string {}, with rest {}", input, rest)
+            }
+            ParseError::NomError(err) => write!(f, "Parsing failed with NomError: {}", err),
+        }
+    }
+}
+
+pub fn house_number(input: &str) -> IResult<&str, HouseNumber> {
```
Comment (zandemax) [inline src/parser/housenumber.rs]: all the following functions do not need to be pub

Comment (zandemax) [reply src/parser/housenumber.rs]: and yeah, i am aware that i made them pub and should have cleaned that up myself :P

File: src/parser/parsertest.rs
Comment (zandemax) [inline src/parser/parsertest.rs:1]: delete

File: src/pbf.rs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    point: Point,
+    text: String,
+}
+
+pub struct Buildings (
+    pub Vec<Building>
+);
+
+impl Buildings {
+    pub fn distribute_population(mut self, housenumbers: Vec<HouseNumberPoint>, inhabitants_total: u64, config: &Config) -> Self {
```
Comment (zandemax) [inline src/pbf.rs]: It feels a bit weird that this function essentially mutates the Struct, but consumes it to give back the same type.
Maybe the type before or after the mutation should be different.

File: src/pbf.rs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        self.0 = self.0.into_iter().map(|building| {
+            let mut building = building;
+            let flat_count = building.tags["flats"].parse::<u32>().unwrap();
+            let mut population = 0;
+            for flat_index in flat_offset..(flat_offset+flat_count) {
+                population += flat_inhabitants[flat_index as usize];
+            }
+            flat_offset += flat_count;
+            building.tags.insert("pop".to_string().into(), population.to_string().into());
+            building
```
Comment (zandemax) [inline src/pbf.rs]: Since this is changing the building, it should be implemented as a method of building that can just be called in the map call.

File: src/pbf.rs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        while inhabitants_to_distribute > 0 {
+            let flat_index = rand::thread_rng().gen_range(0..total_flat_count - 1) as usize;
+            if flat_inhabitants[flat_index] > config.reroll_threshold
+                && rand::thread_rng().gen_range(0..config.reroll_probability) > config.reroll_threshold
+            {
+                continue;
+            }
+            flat_inhabitants[flat_index] += 1;
+            inhabitants_to_distribute -= 1;
+        }
```
Comment (zandemax) [inline src/pbf.rs]: I don't think that pre-calculating the populations here is necessary. Instead, I think we could determine the probability distribution function for the population distribution across households (e.g. from [Destatis Genesis](https://www-genesis.destatis.de) table 12211-9022) and then sample from that distribution for each household below.
e.g. just assuming a normal distribution with $\sigma$ of 0.3 and calculating the mean:
```Rust
let inhabitants_mean = inhabitants_total/total_flat_count;
let inhabitants_distribution = Normal::new(inhabitants_mean, 0.3).unwrap();

buildings.iter().map(|building| 
    building.flats
                   .iter()
                   .map(|flat| 
                       flat.set_inhabitants(
                           inhabitants_distribution.sample().round()
                    )));
```

File: src/pbf.rs [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        if self.tags.contains_key("addr:housenumber") {
+            house_numbers = match HouseNumberList::try_from(self.tags["addr:housenumber"].as_str())
+            {
+                Ok(house_number) => house_number,
+                Err(err) => {
+                    println!("Error: {:?}", err);
+                    HouseNumberList::new()
+                }
+            };
+        }
```
Comment (zandemax) [inline src/pbf.rs]:
```suggestion
        let mut house_numbers = if self.tags.contains_key("addr:housenumber") {
            HouseNumberList::try_from(self.tags["addr:housenumber"].as_str()).unwrap_or_default()
        } else {
            HouseNumberList::new()
        }
```

File: src/pbf.rs [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+            }
+        }
+
+        // Insert flat tag
+        self.tags.insert(
+            "flats".to_string().into(),
+            flat_count.to_string().into(),
+        );
+        self
+    }
```
Comment (zandemax) [inline src/pbf.rs]: I think we should split this up into multiple functions, else it is too hard to follow

File: src/pbf.rs (hunk 5 above)
Comment (TheNewCivilian) [reply src/pbf.rs]: Let's add a distribution function, that allows users without knowledge about the exact population to populate buildings using the known German / Austrian inhabitant distribution per flat.

File: src/pbf.rs (hunk 7 above)
Comment (TheNewCivilian) [reply src/pbf.rs]: Its not exactly split up right now.. but more documentation and cleanup happened. Maybe you @zandemax  can check it again, if that already did it..

File: src/pbf.rs (hunk 5 above)
Comment (TheNewCivilian) [reply src/pbf.rs]: I opened an issue for that here: #7

**Review by zandemax — APPROVED** [review]
Looks good to me :)
